### PR TITLE
[clang] Always allow including builtin headers in [no_undeclared_headers] modules.

### PR DESCRIPTION
**Pull request by martinboehme**
Note on cherrypick:
This is required to make the `SwiftGlibc` module map work with libc++.

Original commit message:
Previously, this would fail if the builtin headers had been "claimed" by
a different module that wraps these builtin headers. libc++ does this,
for example.

This change adds a test demonstrating this situation; the test fails
without the fix.

### CODE REVIEW

**Comment by martinboehme** [comment]
@swift-ci Please test

**Comment by martinboehme** [comment]
@dexonsmith Could you find a suitable reviewer? Thanks!

**Comment by hyp** [comment]
That looks pretty good. I'd like to test this internally before merging. I'll get back to you either tomorrow or Monday.

**Comment by martinboehme** [comment]
@hyp Any updates?

**Review by hyp — APPROVED** [review]

**Comment by martinboehme** [comment]
@swift-ci Please test

**Comment by martinboehme** [comment]
@hyp Could you or someone else with write access merge this? (I can't...)

I've triggered another test to make sure this doesn't conflict with anything that was committed in the meantime.

**Pull request merged**
